packages/protocol/contracts/L1/verifiers/SGXVerifier.sol

```diff
@@ -6,39 +6,154 @@
 
 pragma solidity ^0.8.20;
 
+import { ECDSAUpgradeable } from "@ozu/utils/cryptography/ECDSAUpgradeable.sol";
+
 import { EssentialContract } from "../../common/EssentialContract.sol";
 import { Proxied } from "../../common/Proxied.sol";
+import { LibBytesUtils } from "../../thirdparty/LibBytesUtils.sol";
 
 import { TaikoData } from "../TaikoData.sol";
 
 import { IVerifier } from "./IVerifier.sol";
 
-/// @title GuardianVerifier
-// TODO(dani): implement this verifier.
+/// @title SGXVerifier
 contract SGXVerifier is EssentialContract, IVerifier {
-    uint256[50] private __gap;
+    using ECDSAUpgradeable for bytes32;
+
+    uint256 public constant INSTANCE_EXPIRY = 180 days;
+
+    mapping(address instance => uint256 registeredAt) public instances;
+
+    uint256[49] private __gap;
+
+    event InstanceRegistered(
+        address indexed replaced, address indexed instance, uint256 registeredAt
+    );
 
-    error UNIMPLEMENTED();
+    error SGX_INVALID_AUTH();
+    error SGX_INVALID_INSTANCE();
+    error SGX_INVALID_PROOF_SIZE();
+    error SGX_INSTANCE_REGISTERED();
 
     /// @notice Initializes the contract with the provided address manager.
     /// @param _addressManager The address of the address manager contract.
     function init(address _addressManager) external initializer {
         EssentialContract._init(_addressManager);
     }
 
+    /// @notice Adds trusted SGX instances to the registry.
+    /// @param trustedInstances The address array of trusted SGX instances.
+    function registerInstance(address[] memory trustedInstances)
+        external
+        onlyOwner
+    {
+        for (uint256 i; i < trustedInstances.length; i++) {
+            _replaceInstance(address(0), trustedInstances[i]);
+        }
+    }
+
+    /// @notice Removes trusted SGX instances from the registry.
+    /// @param oldInstance The address of a compromised SGX instance.
+    function invalidateInstance(address oldInstance) external onlyOwner {
+        instances[oldInstance] = 1;
+    }
+
+    /// @notice Adds trusted SGX instances to the registry by another SGX
+    /// instance.
+    /// @param newAddress The new address of the instance.
+    /// @param trustedInstances The address array of trusted SGX instances.
+    /// @param signature The signature proving authenticity.
+    function registerBySgxInstance(
+        address newAddress,
+        address[] memory trustedInstances,
```

Comment:
what's the difference between 'newAddress' and 'trustedInstances', why `trustedInstances` cannot include `newAddress`?

Reply:
It can, so that the first (or last) can be the instance's new address, sure, can do it.

Reply:
Logically having them separate makes more sense to me. `trustedInstances` is the list of addresses that an SGX instance has verified the remote attestations of, and `newAddress` is his own new public key for the next SGX proof that SGX instance will generate. Combining them is a bit confusing I feel.